File: php/bench/fold_bench.exs
# php/bench/fold_bench.exs — the Elixir side of the cold/warm fold benchmark.
#
#   mix run php/bench/fold_bench.exs [--iterations N]   (warm)
#   mix run php/bench/fold_bench.exs --cold             (one fold then exit)
#
# Same workload as the PHP bench: load the 422156 fixture + run the full ingest fold to the final
# golden record (GoldenRecords.from_envelopes/2). WARM loads the fixture once then folds N times,
# timing each with System.monotonic_time, and reports median + p99 µs/fold as JSON. COLD folds once
# and exits — run.sh times ~20 fresh `elixir`/`mix run` invocations and averages the wall clock.

fixture = Path.join(__DIR__, "../../test/ingest/fixtures/medipim_be_422156.json")

{opts, _, _} = OptionParser.parse(System.argv(), strict: [cold: :boolean, iterations: :integer])
cold = Keyword.get(opts, :cold, false)
iterations = Keyword.get(opts, :iterations, 2000)

fold = fn env -> GoldenRecords.from_envelopes([env], 1) end

if cold do
  env = HistoryEnvelope.load!(fixture)
  %{records: records} = fold.(env)
  IO.puts(:stderr, "cold fold: #{length(records)} product(s)")
  System.halt(0)
end

env = HistoryEnvelope.load!(fixture)

# warm-up
Enum.each(1..50, fn _ -> fold.(env) end)

samples =
  for _ <- 1..iterations do
    t0 = System.monotonic_time(:nanosecond)
    fold.(env)
    (System.monotonic_time(:nanosecond) - t0) / 1000.0
  end

sorted = Enum.sort(samples)
n = length(sorted)
median = Enum.at(sorted, div(n, 2))
p99 = Enum.at(sorted, min(n - 1, trunc(n * 0.99)))
mean = Enum.sum(sorted) / n

doc = %{
  "lang" => "elixir",
  "mode" => "warm",
  "elixir_version" => System.version(),
  "otp_release" => :erlang.system_info(:otp_release) |> to_string(),
  "iterations" => iterations,
  "median_us" => Float.round(median, 2),
  "p99_us" => Float.round(p99, 2),
  "mean_us" => Float.round(mean, 2),
  "folds_per_sec" => Float.round(1_000_000 / median, 1)
}

IO.puts(JSON.encode!(doc))

File: php/bench/fold_bench.php
<?php

declare(strict_types=1);

/**
 * php/bench/fold_bench.php — the PHP side of the cold/warm fold benchmark.
 *
 * Workload = load the 422156 fixture + run the full ingest fold to the final golden record
 * (GoldenRecords::fromEnvelopes). Two modes:
 *
 *   WARM (default):  load the fixture ONCE, then fold N times (default 2000), timing each fold with
 *                    hrtime(true). Reports median + p99 µs/fold and folds/sec as JSON.
 *   COLD (--cold):   one fold, then exit — so a fresh `php -d opcache.enable_cli=1` invocation times
 *                    the whole interpreter+autoload+fold path. run.sh averages ~20 such invocations.
 *
 * Usage:
 *   php -d opcache.enable_cli=1 php/bench/fold_bench.php [--iterations=2000]
 *   php -d opcache.enable_cli=1 php/bench/fold_bench.php --cold
 */

require __DIR__.'/../vendor/autoload.php';

use GoldenRecord\EnvelopeLoader;
use GoldenRecord\GoldenRecords;

$fixture = __DIR__.'/../../test/ingest/fixtures/medipim_be_422156.json';

$argvOpts = getopt('', ['cold', 'iterations::']);
$cold = isset($argvOpts['cold']);
$iterations = isset($argvOpts['iterations']) ? max(1, (int) $argvOpts['iterations']) : 2000;

/** One full fold: parse the already-read envelope + project to the golden record. */
$fold = static function (array $env): array {
    return GoldenRecords::fromEnvelopes([$env], 1);
};

if ($cold) {
    // COLD: the timed unit is this whole process. We still parse + fold once so the work is real;
    // the wall-clock is measured by `time` around the invocation in run.sh.
    $env = EnvelopeLoader::loadBang($fixture);
    $records = $fold($env)['records'];
    // touch the result so nothing is optimized away
    fwrite(STDERR, 'cold fold: '.count($records)." product(s)\n");
    exit(0);
}

// WARM: load the fixture once, then fold N times, timing each.
$env = EnvelopeLoader::loadBang($fixture);

// warm-up (let opcache settle, prime any lazy state)
for ($i = 0; $i < 50; ++$i) {
    $fold($env);
}

$samples = [];
for ($i = 0; $i < $iterations; ++$i) {
    $t0 = hrtime(true);
    $fold($env);
    $samples[] = (hrtime(true) - $t0) / 1000.0; // nanoseconds -> microseconds
}

sort($samples);
$median = $samples[intdiv(count($samples), 2)];
$p99 = $samples[(int) floor(count($samples) * 0.99)];
$mean = array_sum($samples) / count($samples);

echo json_encode([
    'lang' => 'php',
    'mode' => 'warm',
    'php_version' => PHP_VERSION,
    'iterations' => $iterations,
    'median_us' => round($median, 2),
    'p99_us' => round($p99, 2),
    'mean_us' => round($mean, 2),
    'folds_per_sec' => round(1_000_000 / $median, 1),
], JSON_PRETTY_PRINT).PHP_EOL;
